### PR TITLE
gh-104050: Improve Argument Clinic type annotation coverage

### PR DESCRIPTION
**Pull request by erlend-aasland**
Add various missing annotations in the following classes:
- BlockPrinter
- CConverter
- CLanguage
- FormatCounterFormatter
- Language
- _TextAccumulator

* Issue: gh-104050

### CODE REVIEW

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -432,10 +432,11 @@ class FormatCounterFormatter(string.Formatter):
          the counts dict would now look like
          {'a': 2, 'b': 1, 'c': 1}
     """
-    def __init__(self):
-        self.counts = collections.Counter()
+    def __init__(self) -> None:
+        self.counts: dict[str, int] = collections.Counter()
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: I would go for

```suggestion
        self.counts = collections.Counter[str]()
```

Or

```suggestion
        self.counts: collections.Counter[str] = collections.Counter()
```

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -3000,7 +3014,7 @@ def pre_render(self):
         """
         pass
 
-    def parse_arg(self, argname, displayname):
+    def parse_arg(self, argname: str, displayname: str):
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:3016]:
```suggestion
    def parse_arg(self, argname: str, displayname: str) -> str:
```

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -432,10 +432,11 @@ class FormatCounterFormatter(string.Formatter):
...
          {'a': 2, 'b': 1, 'c': 1}
     """
-    def __init__(self):
-        self.counts = collections.Counter()
+    def __init__(self) -> None:
+        self.counts: dict[str, int] = collections.Counter()
 
-    def get_value(self, key, args, kwargs):
+    def get_value(self, key: str | int, args, kwargs) -> str:
+        assert isinstance(key, str)
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: I would just do:

```suggestion
    def get_value(self, key: str, args, kwargs) -> str:  # type: ignore[override]
```

Mypy's complaining that we're overriding the superclass's method in a way that violates the Liskov Substitution Principle. But that's okay, we know what we're doing here; it's fine in this instance to just tell mypy to go away.

File: Tools/clinic/clinic.py [hunk 4]
```diff
@@ -748,10 +756,10 @@ class CLanguage(Language):
     stop_line     = "[{dsl_name} start generated code]*/"
     checksum_line = "/*[{dsl_name} end generated code: {arguments}]*/"
 
-    def __init__(self, filename):
+    def __init__(self, filename: str) -> None:
         super().__init__(filename)
         self.cpp = cpp.Monitor(filename)
-        self.cpp.fail = fail
+        setattr(self.cpp, "fail", fail)
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: Here again, I'd just leave it as it is and tell mypy to go away:

```suggestion
        self.cpp.fail = fail  # type: ignore[method-assign]
```

Mypy doesn't really understand what we're doing either way, so better just to keep the code idiomatic Python imo

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Oh, TIL!

File: Tools/clinic/clinic.py (hunk 2 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:3016]: IIRC, this caused a lot of havoc, so I'd rather save it for later :)

Comment (AlexWaygood) [reply Tools/clinic/clinic.py:3016]: Oh, sure!

**Review by AlexWaygood — APPROVED** [review]
Thanks!

**Pull request merged**